File: js/scoring.js
// ═══════════════════════════════════════════════════════════════
// AgroXitique — Motor de Scoring (Programação Funcional)
// Implementação em JavaScript usando princípios funcionais puros
// ═══════════════════════════════════════════════════════════════

const AgroScoring = (() => {
    // ── Funções utilitárias puras ──────────────────────────────
    const clamp = (min, max) => (value) =>
        Math.max(min, Math.min(max, value));

    const clampScore = clamp(0, 100);

    const pipe = (...fns) => (x) =>
        fns.reduce((acc, fn) => fn(acc), x);

    const sum = (arr) => arr.reduce((a, b) => a + b, 0);

    const percentage = (part, total) =>
        total === 0 ? 0 : Math.round((part / total) * 100);

    // ── Factores de scoring (funções puras) ───────────────────

    /**
     * Pontualidade — 40% do score
     * +10 pontos por cada contribuição feita a tempo
     * Normalizado para escala 0-40
     */
    const calcularPontualidade = (contribuicoesPontuais, totalMeses) => {
        if (totalMeses === 0) return 0;
        const ratio = contribuicoesPontuais / totalMeses;
        return Math.round(ratio * 40);
    };

    /**
     * Consistência — 30% do score
     * Baseada em meses consecutivos activos
     * Normalizado para escala 0-30
     */
    const calcularConsistencia = (mesesActivos, totalMeses) => {
        if (totalMeses === 0) return 0;
        const ratio = mesesActivos / Math.max(totalMeses, 6);
        return Math.round(ratio * 30);
    };

    /**
     * Tamanho do grupo — 10% do score
     * Bónus por grupos maiores e mais estáveis
     */
    const calcularBonusGrupo = (numMembros) => {
        if (numMembros >= 12) return 10;
        if (numMembros >= 8) return 7;
        if (numMembros >= 5) return 5;
        return 2;
    };

    /**
     * Histórico de crédito — 15% do score
     * +5 pontos por cada reembolso bem-sucedido, max 15
     */
    const calcularHistoricoCredito = (reembolsosFeitos) => {
        return Math.min(15, reembolsosFeitos * 5);
    };

    /**
     * Indicações — 5% do score
     * +2.5 pontos por cada membro indicado, max 5
     */
    const calcularIndicacoes = (indicacoesFeitas) => {
        return Math.min(5, indicacoesFeitas * 2.5);
    };

    /**
     * Penalidades
     * -15 por atraso de pagamento
     * -20 por falta de contribuição num mês
     */
    const calcularPenalidades = (atrasos, faltas) => {
        return (atrasos * -15) + (faltas * -20);
    };

    // ── Composição final do score ─────────────────────────────

    /**
     * calcularScoreIndividual :: UserData -> ScoreResult
     * Função pura: mesmo input = mesmo output (determinística)
     */
    const calcularScoreIndividual = (userData) => {
        const pontualidade = calcularPontualidade(
            userData.contribuicoesPontuais || userData.monthsActive,
            userData.totalMeses || 6
        );
        const consistencia = calcularConsistencia(
            userData.monthsActive || 0,
            userData.totalMeses || 6
        );
        const bonusGrupo = calcularBonusGrupo(userData.numMembrosGrupo || 0);
        const historicoCredito = calcularHistoricoCredito(userData.reembolsosFeitos || 0);
        const indicacoes = calcularIndicacoes(userData.indicacoesFeitas || 0);
        const penalidades = calcularPenalidades(
            userData.atrasos || 0,
            userData.faltas || 0
        );

        const scoreRaw = pontualidade + consistencia + bonusGrupo +
                         historicoCredito + indicacoes + penalidades;

        const scoreFinal = clampScore(scoreRaw);

        return {
            total: scoreFinal,
            classificacao: classificarScore(scoreFinal),
            factores: {
                pontualidade,
                consistencia,
                bonusGrupo,
                historicoCredito,
                indicacoes,
                penalidades,
            },
            limiteCredito: calcularLimiteCredito(scoreFinal),
            nivelRisco: avaliarRisco(scoreFinal),
        };
    };

    /**
     * calcularScoreGrupo :: [MemberData] -> GroupScoreResult
     * Score do grupo é a média ponderada dos scores individuais
     */
    const calcularScoreGrupo = (members) => {
        if (members.length === 0) return { total: 0, classificacao: "Sem dados" };

        const scores = members.map(m => m.score || 0);
        const media = Math.round(sum(scores) / scores.length);
        const membrosEmDia = members.filter(m => m.paid).length;
        const taxaPagamento = percentage(membrosEmDia, members.length);

        // Ajuste: penalizar se muitos membros em atraso
        const ajuste = taxaPagamento >= 75 ? 5 : taxaPagamento >= 50 ? 0 : -10;
        const scoreFinal = clampScore(media + ajuste);

        return {
            total: scoreFinal,
            classificacao: classificarScore(scoreFinal),
            membrosEmDia,
            totalMembros: members.length,
            taxaPagamento,
            limiteCredito: calcularLimiteCreditoGrupo(scoreFinal, members.length),
        };
    };

    // ── Classificação ─────────────────────────────────────────

    const classificarScore = (score) => {
        if (score >= 85) return "Excelente";
        if (score >= 70) return "Bom";
        if (score >= 50) return "Regular";
        if (score >= 30) return "Fraco";
        return "Crítico";
    };

    const classificarScoreEmoji = (score) => {
        if (score >= 85) return "🟢";
        if (score >= 70) return "🟡";
        if (score >= 50) return "🟠";
        return "🔴";
    };

    // ── Cálculo de crédito ────────────────────────────────────

    const calcularLimiteCredito = (score) => {
        if (score >= 85) return 10000;
        if (score >= 70) return 8000;
        if (score >= 50) return 4000;
        if (score >= 30) return 2000;
        return 0;
    };

    const calcularLimiteCreditoGrupo = (score, numMembros) => {
        const base = calcularLimiteCredito(score);
        const multiplicador = Math.min(numMembros / 5, 2.0);
        return Math.round(base * multiplicador);
    };

    const calcularParcela = (valor, taxaMensal, numParcelas) => {
        // Juros simples (mais justo para micro-crédito rural)
        const jurosTotal = valor * (taxaMensal / 100) * numParcelas;
        return Math.round((valor + jurosTotal) / numParcelas);
    };

    // ── Avaliação de risco ────────────────────────────────────

    const avaliarRisco = (score) => {
        if (score >= 75) return { nivel: "Baixo", cor: "low", probabilidadeDefault: 5 };
        if (score >= 50) return { nivel: "Médio", cor: "medium", probabilidadeDefault: 15 };
        return { nivel: "Alto", cor: "high", probabilidadeDefault: 35 };
    };

    const simularCenarios = (grupoScore, valorCredito, numMembros) => {
        return [
            {
                nome: "Optimista",
                descricao: "Todos pagam a tempo, boa colheita",
                probabilidade: grupoScore >= 70 ? 60 : 30,
                impactoScore: +10,
            },
            {
                nome: "Base",
                descricao: "Maioria paga, colheita normal",
                probabilidade: 30,
                impactoScore: 0,
            },
            {
                nome: "Pessimista",
                descricao: "Atrasos frequentes, má colheita",
                probabilidade: grupoScore >= 70 ? 10 : 40,
                impactoScore: -15,
            },
        ];
    };

    // ── API pública ───────────────────────────────────────────
    return {
        calcularScoreIndividual,
        calcularScoreGrupo,
        classificarScore,
        classificarScoreEmoji,
        calcularLimiteCredito,
        calcularLimiteCreditoGrupo,
        calcularParcela,
        avaliarRisco,
        simularCenarios,
        calcularPontualidade,
        calcularConsistencia,
        calcularBonusGrupo,
        calcularPenalidades,
        clampScore,
        pipe,
    };
})();

if (typeof window !== 'undefined') {
    window.AgroScoring = AgroScoring;
}

File: haskell/Scoring.hs
{- ═══════════════════════════════════════════════════════════════
   AgroXitique — Motor de Scoring
   Funções puras para cálculo do score de crédito rural
   
   Princípios:
   - Funções puras (mesmo input → mesmo output)
   - Composabilidade (cada factor é uma função independente)
   - Testabilidade (cada função testável isoladamente no GHCi)
   
   Cadeira: Programação Funcional — UniLicungo 2025
   Professor: Filipe Domingos dos Santos
   ═══════════════════════════════════════════════════════════════ -}

module Scoring where

import Types

-- ═══════════════════════════════════════════════════════════════
-- Funções auxiliares puras
-- ═══════════════════════════════════════════════════════════════

-- | Limita um valor entre mínimo e máximo
--   clamp 0 100 120 == 100
--   clamp 0 100 (-5) == 0
clamp :: Int -> Int -> Int -> Int
clamp minVal maxVal x = max minVal (min maxVal x)

-- | Arredonda uma percentagem para inteiro
percentagem :: Int -> Int -> Int
percentagem parte total
    | total == 0 = 0
    | otherwise  = round (fromIntegral parte / fromIntegral total * 100 :: Double)

-- ═══════════════════════════════════════════════════════════════
-- Factores individuais do score (cada um é uma função pura)
-- ═══════════════════════════════════════════════════════════════

-- | Factor 1: Pontualidade (peso: 40%)
--   Conta contribuições feitas a tempo e normaliza para 0-40
--
--   Exemplo:
--   > calcPontualidade 5 6
--   33
calcPontualidade :: Int    -- ^ Contribuições pontuais
                 -> Int    -- ^ Total de meses
                 -> Int    -- ^ Pontuação (0-40)
calcPontualidade pontuais totalMeses
    | totalMeses == 0 = 0
    | otherwise       = round (fromIntegral pontuais / fromIntegral totalMeses * 40 :: Double)

-- | Factor 2: Consistência (peso: 30%)
--   Baseada em meses consecutivos activos, normalizado por 6 meses
--
--   Exemplo:
--   > calcConsistencia 4 6
--   20
calcConsistencia :: Int    -- ^ Meses activos
                 -> Int    -- ^ Total de meses (mínimo referência: 6)
                 -> Int    -- ^ Pontuação (0-30)
calcConsistencia mesesActivos totalMeses
    | totalMeses == 0 = 0
    | otherwise       = round (fromIntegral mesesActivos / fromIntegral (max totalMeses 6) * 30 :: Double)

-- | Factor 3: Bónus por tamanho do grupo (peso: 10%)
--   Grupos maiores têm mais estabilidade
--
--   Exemplo:
--   > calcBonusGrupo 12
--   10
--   > calcBonusGrupo 3
--   2
calcBonusGrupo :: Int    -- ^ Número de membros
               -> Int    -- ^ Pontuação (0-10)
calcBonusGrupo n
    | n >= 12   = 10
    | n >= 8    = 7
    | n >= 5    = 5
    | otherwise = 2

-- | Factor 4: Histórico de crédito (peso: 15%)
--   +5 pontos por cada reembolso bem-sucedido
--
--   Exemplo:
--   > calcHistoricoCredito 3
--   15
calcHistoricoCredito :: Int    -- ^ Reembolsos feitos
                     -> Int    -- ^ Pontuação (0-15)
calcHistoricoCredito reembolsos = min 15 (reembolsos * 5)

-- | Factor 5: Indicações (peso: 5%)
--   +2 pontos por cada membro indicado
--
--   Exemplo:
--   > calcIndicacoes 2
--   4
calcIndicacoes :: Int    -- ^ Indicações feitas
               -> Int    -- ^ Pontuação (0-5)
calcIndicacoes indicacoes = min 5 (indicacoes * 2)

-- | Penalidades
--   -15 por atraso, -20 por falta
--
--   Exemplo:
--   > calcPenalidades 1 0
--   -15
calcPenalidades :: Int    -- ^ Número de atrasos
                -> Int    -- ^ Número de faltas
                -> Int    -- ^ Penalidade (negativo)
calcPenalidades atrasos faltas = (atrasos * (-15)) + (faltas * (-20))


-- ═══════════════════════════════════════════════════════════════
-- Composição do score final
-- ═══════════════════════════════════════════════════════════════

-- | Calcula o score individual de um membro
--   Combina todos os factores numa pontuação final de 0 a 100
--
--   Exemplo:
--   > let m = Membro "Maria" "84123" Activo [Contribuicao 200 True (MesAno 5 2025)] 5 1
--   > calcularScoreIndividual m 8
--   ScoreResult {scoreTotal = 72, ...}
calcularScoreIndividual :: Membro -> Int -> ScoreResult
calcularScoreIndividual membro numMembrosGrupo =
    let contribs   = memContribs membro
        pontuais   = length (filter contEmDia contribs)
        totalMeses = max (length contribs) (memMesesActivos membro)
        
        -- Calcular cada factor independentemente
        fPontualidade = calcPontualidade pontuais totalMeses
        fConsistencia = calcConsistencia (memMesesActivos membro) totalMeses
        fBonusGrupo   = calcBonusGrupo numMembrosGrupo
        fHistCredito  = calcHistoricoCredito 0  -- Sem empréstimos anteriores
        fIndicacoes   = calcIndicacoes (memIndicacoes membro)
        fPenalidades  = calcPenalidades (contarAtrasos contribs) 0
        
        -- Composição: soma de todos os factores
        scoreRaw = fPontualidade + fConsistencia + fBonusGrupo 
                 + fHistCredito + fIndicacoes + fPenalidades
        
        scoreFinal = clamp 0 100 scoreRaw
        
        factores = ScoreFactores
            { factPontualidade     = fPontualidade
            , factConsistencia     = fConsistencia
            , factBonusGrupo       = fBonusGrupo
            , factHistoricoCredito = fHistCredito
            , factIndicacoes       = fIndicacoes
            , factPenalidades      = fPenalidades
            }
    in ScoreResult
        { scoreTotal         = scoreFinal
        , scoreClassificacao = classificar scoreFinal
        , scoreLimiteCredito = calcularLimiteCredito scoreFinal
        , scoreFactores      = factores
        }

-- | Conta contribuições em atraso (não pontuais)
contarAtrasos :: [Contribuicao] -> Int
contarAtrasos = length . filter (not . contEmDia)

-- | Calcula o score médio de um grupo
--   O score do grupo é a média dos scores individuais, com ajuste
calcularScoreGrupo :: Grupo -> Int
calcularScoreGrupo grupo =
    let membros = grupoMembros grupo
        n       = length membros
        scores  = map (\m -> calcularScoreIndividual m n) membros
        soma    = sum (map scoreTotal scores)
        media   = if n == 0 then 0 else soma `div` n
        
        -- Ajuste: penalizar se muitos pendentes
        pendentes    = length (filter (\m -> memStatus m == Pendente) membros)
        taxaPagamento = if n == 0 then 0 else percentagem (n - pendentes) n
        ajuste       = if taxaPagamento >= 75 then 5
                       else if taxaPagamento >= 50 then 0
                       else (-10)
    in clamp 0 100 (media + ajuste)


-- ═══════════════════════════════════════════════════════════════
-- Classificação e limites de crédito
-- ═══════════════════════════════════════════════════════════════

-- | Classifica um score numérico
--
--   > classificar 85
--   Excelente
--   > classificar 72
--   Bom
classificar :: Int -> Classificacao
classificar s
    | s >= 85   = Excelente
    | s >= 70   = Bom
    | s >= 50   = Regular
    | s >= 30   = Fraco
    | otherwise = Critico

-- | Determina o limite de crédito baseado no score
--
--   > calcularLimiteCredito 85
--   10000.0
--   > calcularLimiteCredito 50
--   4000.0
calcularLimiteCredito :: Int -> Double
calcularLimiteCredito s
    | s >= 85   = 10000.0
    | s >= 70   = 8000.0
    | s >= 50   = 4000.0
    | s >= 30   = 2000.0
    | otherwise = 0.0

-- | Calcula o limite de crédito do grupo (ajustado por nº de membros)
calcularLimiteCreditoGrupo :: Int -> Int -> Double
calcularLimiteCreditoGrupo score numMembros =
    let base          = calcularLimiteCredito score
        multiplicador = min (fromIntegral numMembros / 5.0) 2.0 :: Double
    in base * multiplicador

-- | Calcula o valor de cada parcela (juros simples)
--
--   > calcularParcela 4000 3 4
--   1060.0
calcularParcela :: Double  -- ^ Valor do crédito
                -> Double  -- ^ Taxa mensal (%)
                -> Int     -- ^ Número de parcelas
                -> Double  -- ^ Valor de cada parcela
calcularParcela valor taxaMensal numParcelas =
    let jurosTotal = valor * (taxaMensal / 100) * fromIntegral numParcelas
    in (valor + jurosTotal) / fromIntegral numParcelas


-- ═══════════════════════════════════════════════════════════════
-- Funções de conveniência para demonstração
-- ═══════════════════════════════════════════════════════════════

-- | Cria uma contribuição pontual
contribPontual :: Double -> MesAno -> Contribuicao
contribPontual val mesAno = Contribuicao val True mesAno

-- | Cria uma contribuição em atraso
contribAtrasada :: Double -> MesAno -> Contribuicao
contribAtrasada val mesAno = Contribuicao val False mesAno

-- | Mostra o score de forma legível
mostrarScore :: ScoreResult -> String
mostrarScore sr = unlines
    [ "══════════════════════════════════"
    , "  Score: " ++ show (scoreTotal sr) ++ "/100 — " ++ show (scoreClassificacao sr)
    , "  Limite de crédito: " ++ show (scoreLimiteCredito sr) ++ " MT"
    , "──────────────────────────────────"
    , "  Factores:"
    , "    Pontualidade:     +" ++ show (factPontualidade f)
    , "    Consistência:     +" ++ show (factConsistencia f)
    , "    Bónus grupo:      +" ++ show (factBonusGrupo f)
    , "    Hist. crédito:    +" ++ show (factHistoricoCredito f)
    , "    Indicações:       +" ++ show (factIndicacoes f)
    , "    Penalidades:      " ++ show (factPenalidades f)
    , "══════════════════════════════════"
    ]
  where f = scoreFactores sr
